ethcore/src/engines/tendermint/mod.rs

```diff
@@ -590,6 +637,24 @@ impl Engine for Tendermint {
 		None
 	}
 
+	fn epoch_verifier<'a>(&self, _header: &Header, proof: &'a [u8]) -> ConstructedVerifier<'a> {
+		let (signal_number, set_proof) = match destructure_proofs(proof) {
+			Ok(x) => x,
+			Err(e) => return ConstructedVerifier::Err(e),
+		};
+
+		let first = signal_number == 0;
+		match self.validators.epoch_set(first, self, signal_number, set_proof) {
+			Ok((list, _)) => {
+				// Tendermint blocks have instant finality, so no need to check finality proof
+				ConstructedVerifier::Trusted(Box::new(EpochVerifier {
```

Comment:
it should decode correctly into the two portions (signal proof and finality proof), and each portion should be checked. signal proof is checked if `epoch_set` returns `Ok`, and `finality_proof` will be checked at that point long as we return `ConstructedVerifier::Unconfirmed` from `epoch_verifier`